vulnerabilities/unsecure-signatures.md

```diff
@@ -1,2 +1,82 @@
 ## Unsecure Signatures
 
+### Overview
+
+`ecrecover` is a Solidity function used to recover the address associated with an elliptic curve signature. Improper use can lead to significant vulnerabilities, including null address recovery, signature malleability, and signature replay attacks.
+
+### Vulnerability Example
+
+Consider the following function that allows users to change ownership based on a signature:
+
+```solidity
+// VULNERABLE CODE
+function changeOwner(bytes32 ownerHash, uint8 v, bytes32 r, bytes32 s) external {
+    address signer = ecrecover(ownerHash, v, r, s);
+    require(signer == owner, "Invalid signature");
+    owner = ownerHash;
+}
+```
+
+**Problems**:
+
+1. **Null Address Recovery**: If `v` is set to any value other than 27 or 28, `ecrecover` can return `address(0)`, passing the `require` check if `owner` is also `address(0)`.
+2. **Signature Malleability**: Attackers can modify `r` and `s` to create alternate valid signatures for the same message.
+3. **Signature Replay**: Previously used signatures can be reused, allowing unauthorized actions on the contract.
+
+### Mitigation: EIP-712 & OpenZeppelin’s ECDSA
+
+To address these vulnerabilities, use the EIP-712 standard for signing structured data along with OpenZeppelin’s ECDSA library. This combination ensures signatures are unique, context-specific, and mitigates replay attacks and malleability.
+
+### Secure Code Example
+
+```solidity
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+
+contract SecureOwnership {
+    using ECDSA for bytes32;
+
+    address public owner;
+    mapping(bytes => bool) private usedSignatures;
+
+    bytes32 public constant DOMAIN_SEPARATOR = keccak256(abi.encode(
+        keccak256("EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"),
+        keccak256(bytes("SecureOwnership")),
+        keccak256(bytes("1")),
+        block.chainid,
+        address(this)
+    ));
+
+    function changeOwner(address signer, bytes memory signature) external {
+        bytes32 digest = getDigest(signer);
+        address recoveredSigner = digest.recover(signature);
+
+        require(recoveredSigner == signer && recoveredSigner != address(0) && !usedSignatures[signature], "Invalid signature");
+        usedSignatures[signature] = true; // Mark the signature as used
+        owner = signer;
+    }
```

Comment:
I think this function is actually vulnerable to [signature malleability](https://github.com/kadenzipfel/smart-contract-vulnerabilities/blob/master/vulnerabilities/signature-malleability.md). Tbh I don't think the `changeOwner` example for this vulnerability works very well, probably makes more sense to perform some other action executed using a signature

Reply:
Hey, thanks for answering! Using ECDSA library from OpenZeppelin will remove this vulnerability.

Here is a snippet of the `tryRecover` function withing ECDSA:
```javascript
// EIP-2 still allows signature malleability for ecrecover(). Remove this possibility and make the signature
        // unique. Appendix F in the Ethereum Yellow paper (https://ethereum.github.io/yellowpaper/paper.pdf), defines
        // the valid range for s in (301): 0 < s < secp256k1n ÷ 2 + 1, and for v in (302): v ∈ {27, 28}. Most
        // signatures from current libraries generate a unique signature with an s-value in the lower half order.
        //
        // If your library generates malleable signatures, such as s-values in the upper range, calculate a new s-value
        // with 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141 - s1 and flip v from 27 to 28 or
        // vice versa. If your library also generates signatures with 0/1 for v instead 27/28, add 27 to v to accept
        // these malleable signatures as well.
        if (uint256(s) > 0x7FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF5D576E7357A4501DDFE92F46681B20A0) {
            return (address(0), RecoverError.InvalidSignatureS, s);
        } 
```